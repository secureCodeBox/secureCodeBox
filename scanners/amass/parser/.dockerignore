e2e
helpers.js
