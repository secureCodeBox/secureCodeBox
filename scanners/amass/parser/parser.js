// SPDX-FileCopyrightText: 2020 iteratec GmbH
//
// SPDX-License-Identifier: Apache-2.0

async function parse(fileContent) {
  let identifiedDomains = [];

  if (typeof fileContent === "string") {
    identifiedDomains = fileContent
      .split("\n")
      .filter(Boolean)
      .map((domainJson) => JSON.parse(domainJson));
  } else if (typeof fileContent === "object") {
    // If amass identifies a single result it will be automatically parsed as a json object by the sdk & underlying http lib (axios)
    identifiedDomains = [fileContent];
  }

  return identifiedDomains.map((domain) => {
    return {
      name: domain.name,
      description: `Found subdomain ${domain.name}`,
      category: "Subdomain",
      location: domain.name,
      osi_layer: "NETWORK",
      severity: "INFORMATIONAL",
      attributes: {
        tag: domain.tag,
        name: domain.name,
        source: domain.source,
        domain: domain.domain,
        addresses: domain.addresses,
      },
    };
  });
}

module.exports.parse = parse;
