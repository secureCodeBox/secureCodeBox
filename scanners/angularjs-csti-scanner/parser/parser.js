// SPDX-FileCopyrightText: 2020 iteratec GmbH
//
// SPDX-License-Identifier: Apache-2.0

async function parse(fileContent) {

  if (fileContent.length === 0) {
    return []
  }

  const lines = fileContent.split('\n')
  lines.splice(-1, 1)

  return lines.map(line => {
    const method = parseMethod(line)
    const url = parseUrl(line)
    const data = parseData(line, method.length)
    let parameter = null
    let injectedTemplate = null
    if (data.length > 5) {
      parameter = data.substring(0, data.indexOf('='))
      injectedTemplate = data.substring(data.indexOf('=') + 1)
    }

    return {
      name: 'AngularJS template injection',
      description: 'The given in this finding URL is vulnerable to AngularJS template injection which can lead to XSS',
      osi_layer: 'APPLICATION',
      severity: 'HIGH',
      category: 'Template Injection',
      attributes: {
        url: url,
        method: method,
        parameter: parameter,
        injectedTemplate: injectedTemplate,
      }
    }
  })
}

function parseMethod(line) {
  return line.substring(0, line.indexOf('('))
}

function parseUrl(line) {
  console.log(line)
  return line.match(/http[s]?:\/\/[^\n ]+/g)[0]
}

function parseData(line, start) {
  return line.substring(start + 1, line.indexOf('): '))
}


module.exports.parse = parse
