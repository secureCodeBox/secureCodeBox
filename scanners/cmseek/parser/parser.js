// SPDX-FileCopyrightText: the secureCodeBox authors
//
// SPDX-License-Identifier: Apache-2.0

async function parse(findings) {
  let results = []
  // Making sure the CMS is Joomla
  if (findings.cms_id != "joom") {
    return results
  }
  // Check if debug mode is enabled ; if yes add finding
  let parsed_debug_mode_enabled = []

  // I ran into an issue where the time coverted to ISO String was dependant from the timezone of the machine running the test. 
  // This means that if GitHub Actions CI time and local time are different the test will fail.
  // To fix this we need to enforce the timezone in the date string. 
  // cmseek uses the timezone of the machine running the scan, so it will be different machine to machine (or cloud service).
  // https://github.com/Tuhinshubhra/CMSeeK/blob/ce085fee1b5f48db7412911e399bb2c771e73a0f/cmseekdb/basic.py#L296
  // For simplicity UTC time is enforced, and that is by adding a Z to the end of the date string.
  const last_scanned = new Date(findings.last_scanned + "Z").toISOString();
  if (findings.joomla_debug_mode == "enabled") {
    parsed_debug_mode_enabled = {
      name: "Debug mode",
      identified_at: last_scanned,
      description: `Debug mode is enabled on the site`,
      category: "Security Misconfiguration",
      location: findings.url,
      osi_layer: "APPLICATION",
      severity: "MEDIUM",
    }
  }

  // Check if backup files are open; if yes add finding
  let parsed_backupFiles = []
  if ("joomla_backup_files" in findings) {
    parsed_backupFiles = {
      name: "Backup files",
      identified_at: last_scanned,
      description: `Visible Backup files found`,
      category: "Visible internal files",
      location: findings.url,
      osi_layer: "APPLICATION",
      severity: "INFORMATIONAL",
      attributes: {
        joomla_backup_files: findings.joomla_backup_files
      }
    }
  }
  // Check if any core vulnerabilities exist; if yes list findings
  let parsed_vulnerabilities = []
  if (findings.vulnerabilities_count > 0) {
    parsed_vulnerabilities = findings.vulnerabilities.map((vuln) => {
      return {
        name: vuln.name,
        identified_at: last_scanned,
        description: `Vulnerability of type ${vuln.name} found`,
        category: "Vulnerability",
        location: findings.url,
        osi_layer: "APPLICATION",
        severity: "HIGH",
        attributes: {
          joomla_version: findings.joomla_version,
          references: vuln.references,
        }
      };
    });
  }
  // concat all parsed results
  return parsed_vulnerabilities.concat(parsed_backupFiles).concat(parsed_debug_mode_enabled)
}
module.exports.parse = parse;
