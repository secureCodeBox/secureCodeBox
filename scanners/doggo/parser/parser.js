// SPDX-FileCopyrightText: the secureCodeBox authors
//
// SPDX-License-Identifier: Apache-2.0

async function parse(fileContent) {
  const targets = parseResultFile(fileContent);

  if (process.env["DEBUG"] === "true") {
    console.log("Parsing Result File");
    console.log(targets);
  }

  const result = transformToFindings(targets);

  if (process.env["DEBUG"] === "true") {
    console.log("Transform To Findings");
    console.log(result);
  }

  return result;
}

function transformToFindings(targets) {
  // Code to transform the scanner results to scb findings
  return targets.map((target) => {
    return {
      name: `DNS Record: ${target.name} | Type: ${target.type}`,
      description: `DNS record type "${target.type}" found for "${target.name}" with address "${target.address}".`,
      category: "DNS Information",
      location: target.name.slice(0, target.name.length - 1),
      osi_layer: "NETWORK",
      severity: "INFORMATIONAL",
      attributes: {
        doggy_dns_type: target.type,
        ip_addresses: target.type === "A" || target.type === "AAAA" ? [target.address] : undefined,
        doggy_dns_address: target.address,
        doggy_dns_status: target.status,
        doggy_dns_rtt: target.rtt,
        doggy_dns_nameserver: target.nameserver,
      },
    };
  });
}

/**
 * Parses a given doggo result file and extracts all targets
 * @param {*} fileContent
 */
function parseResultFile(fileContent) {
  return Array.isArray(fileContent) ?
    fileContent
    .filter((rawTarget) => {
      // filter out empty targets (domain names which could not be resolved at all)
      return (
        Object.keys(rawTarget).length > 0 &&
        rawTarget.answers !== null &&
        rawTarget.answers !== undefined &&
        rawTarget.answers.length > 0
      );
    })
    .flatMap((rawTarget) => rawTarget.answers) // flatten the answers into one big array
    .filter((rawAnswers) => {
      // filter out empty answers (records which did not exists for individual domains)
      return (
        Object.keys(rawAnswers).length > 0 &&
        rawAnswers.name !== null &&
        rawAnswers.name !== undefined
      );
    }) : [];

}

module.exports.parse = parse;
