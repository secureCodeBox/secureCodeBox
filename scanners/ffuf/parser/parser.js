// SPDX-FileCopyrightText: the secureCodeBox authors
//
// SPDX-License-Identifier: Apache-2.0

async function parse(fileContent) {

  if (!fileContent || !fileContent.results || fileContent.results.length == 0) {
    return [];
  }
  return fileContent.results.map(result => {
    const time = new Date(fileContent.time).toISOString();
    return {
      name: 'Webserver Content',
      description: `Content [${result.input ? Object.values(result.input) : ""}] was found on the webserver ${result.host}.`, // todo rn: what if no FUZZ keyword is used??
      identified_at: time,
      osi_layer: 'APPLICATION',
      severity: 'INFORMATIONAL',
      category: 'Webserver Content',
      attributes: {
        httpStatus: result.status,
        length: result.length,
        words: result.words,
        lines: result.lines,
        contentType: result["content-type"],
        redirectlocation: result.redirectlocation,
        duration: result.duration,
        // resultFile = the name of the file containing the full request and response,
        // SCB does currently not implement saving the file (because data might be large)
        // resultFile: result.resultfile,
        hostname: result.host,
        input: result.input,
        // FUZZ keywords can also be in headers -> we should see that within the result
        postdata: fileContent?.config?.postdata,
        // FUZZ keywords can also be in headers -> we should see that within the result
        headers: fileContent?.config?.headers,
      },
      location: result.url,
    }
  });
}

module.exports.parse = parse;
