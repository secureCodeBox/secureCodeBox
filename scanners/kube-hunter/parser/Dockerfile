# SPDX-FileCopyrightText: 2020 iteratec GmbH
#
# SPDX-License-Identifier: Apache-2.0

ARG namespace
ARG baseImageTag
FROM ${namespace:-securecodebox}/parser-sdk-nodejs:${baseImageTag:-latest}
WORKDIR /home/app/parser-wrapper/parser/
COPY --chown=app:app ./parser.js ./parser.js
