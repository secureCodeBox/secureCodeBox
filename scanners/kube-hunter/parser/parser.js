// SPDX-FileCopyrightText: 2020 iteratec GmbH
//
// SPDX-License-Identifier: Apache-2.0

async function parse({ vulnerabilities = [], nodes = [] }) {
  return vulnerabilities.map(vulnerability => {
    const reference = {}

    if ( vulnerability.vid !== "None") {
      reference.id = vulnerability.vid
      reference.source = `https://aquasecurity.github.io/kube-hunter/kb/${vulnerability.vid}`
    }

    let location = vulnerability.location;
    if (location.startsWith('Local to Pod')) {
      // This is a pod specific vulnarability.
      // As this does not fit the secureCodeBox model to well we will scope this to the first "Node/Master" type node of the cluster.
      // This is subject to change.

      for (const node of nodes) {
        if (node.type === "Node/Master") {
          location = node.location
          break;
        }
      }
    }

    return {
      name: vulnerability.vulnerability,
      description: vulnerability.description,
      location,
      severity: vulnerability.severity.toUpperCase(),
      category: vulnerability.category,
      reference,
      attributes: {
        evidence: vulnerability.evidence,
        kubeHunterRule: vulnerability.hunter,
      }
    };
  });
}

module.exports.parse = parse;
