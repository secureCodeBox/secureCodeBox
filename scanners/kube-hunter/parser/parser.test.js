// SPDX-FileCopyrightText: the secureCodeBox authors
//
// SPDX-License-Identifier: Apache-2.0

import { readFile } from "fs/promises";
import { validateParser } from "@securecodebox/parser-sdk-nodejs/parser-utils";

import { parse } from "./parser";

test("parses result from kind-1.18-in-cluster-scan correctly", async () => {
  const fileContent = JSON.parse(
    await readFile(
      __dirname + "/__testFiles__/kind-1.18-in-cluster-scan.json",
      {
        encoding: "utf8",
      },
    ),
  );
  const findings = await parse(fileContent);
  expect(validateParser(findings)).toBeUndefined();
  expect(findings).toMatchSnapshot();
});

test("should properly parse empty kube-hunter json file", async () => {
  const jsonContent = await readFile(
    __dirname + "/__testFiles__/test-empty-report.json",
    {
      encoding: "utf8",
    },
  );
  const findings = await parse(jsonContent);
  expect(validateParser(findings)).toBeUndefined();
  expect(findings).toMatchInlineSnapshot(`[]`);
});
