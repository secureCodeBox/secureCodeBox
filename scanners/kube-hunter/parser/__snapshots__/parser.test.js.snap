// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parses result from kind-1.18-in-cluster-scan correctly 1`] = `
[
  {
    "attributes": {
      "evidence": "eyJhbGciOiJSUzI1NiIsImtpZCI6IkNabmY2NVgxUmR1ZnQzbHJVQVAzZFFUNjBiR0hUVE9SRDNPcURyenlkODgifQ.eyJpc3MiOiJrdWJlcm5ldGVzL3NlcnZpY2VhY2NvdW50Iiwia3ViZXJuZXRlcy5pby9zZXJ2aWNlYWNjb3VudC9uYW1lc3BhY2UiOiJkZWZhdWx0Iiwia3ViZXJuZXRlcy5pby9zZXJ2aWNlYWNjb3VudC9zZWNyZXQubmFtZSI6Imx1cmNoZXItdG9rZW4tcGpmNGIiLCJrdWJlcm5ldGVzLmlvL3NlcnZpY2VhY2NvdW50L3NlcnZpY2UtYWNjb3VudC5uYW1lIjoibHVyY2hlciIsImt1YmVybmV0ZXMuaW8vc2VydmljZWFjY291bnQvc2VydmljZS1hY2NvdW50LnVpZCI6IjUzOGVhYjdmLTY1YjAtNDE4Yy04MGI2LTI1NGQxNDQ4ODU3NiIsInN1YiI6InN5c3RlbTpzZXJ2aWNlYWNjb3VudDpkZWZhdWx0Omx1cmNoZXIifQ.cGtQHagQ2xxlAFnWwFRNgGJIkaeZIKnqoYYb8GmxN94ry0wwxCbgBm4Kg33A903wDBxd8iuITTk-r8UPZyYJHoxlVu0pHt-3SAc4NT0ob50R2acVXQ2qj_yJOOQHurCWeOJMkGqtCyUoZ8Xcnc6z32Ao-NWzKD-0wV7ndpKm-ytHP0YpHb9bLUPcQGvFoh_UF132yjeJqzwLPRX6hStMYOa8LNhJGyhdejW3BIOylzVPNkKE5lEjWv9f853qnTKG-TzXHBbth7qV8UHwSoY8YFoMezK3zazQt4dN1VG_wYmZ0ujikTC7TRTGr500kFxfpACKwdQ1M1fXgKJhNv9UgA",
      "kubeHunterRule": "Access Secrets",
    },
    "category": "Access Risk",
    "description": " Accessing the pod service account token gives an attacker the option to use the server API ",
    "location": "tcp://10.244.0.1",
    "name": "Read access to pod's service account token",
    "reference": {
      "id": "KHV050",
      "source": "https://aquasecurity.github.io/kube-hunter/kb/KHV050",
    },
    "severity": "LOW",
  },
  {
    "attributes": {
      "evidence": "",
      "kubeHunterRule": "Pod Capabilities Hunter",
    },
    "category": "Access Risk",
    "description": "CAP_NET_RAW is enabled by default for pods. If an attacker manages to compromise a pod, they could potentially take advantage of this capability to perform network attacks on other pods running on the same node",
    "location": "tcp://10.244.0.1",
    "name": "CAP_NET_RAW Enabled",
    "reference": {},
    "severity": "LOW",
  },
  {
    "attributes": {
      "evidence": "['/var/run/secrets/kubernetes.io/serviceaccount/namespace', '/var/run/secrets/kubernetes.io/serviceaccount/ca.crt', '/var/run/secrets/kubernetes.io/serviceaccount/token', '/var/run/secrets/kubernetes.io/serviceaccount/..2020_04_03_14_52_24.460746409/ca.crt', '/var/run/secrets/kubernetes.io/serviceaccount/..2020_04_03_14_52_24.460746409/token', '/var/run/secrets/kubernetes.io/serviceaccount/..2020_04_03_14_52_24.460746409/namespace']",
      "kubeHunterRule": "Access Secrets",
    },
    "category": "Access Risk",
    "description": " Accessing the pod's secrets within a compromised pod might disclose valuable data to a potential attacker",
    "location": "tcp://10.244.0.1",
    "name": "Access to pod's secrets",
    "reference": {},
    "severity": "LOW",
  },
  {
    "attributes": {
      "evidence": "v1.18.0",
      "kubeHunterRule": "Api Version Hunter",
    },
    "category": "Information Disclosure",
    "description": "The kubernetes version could be obtained from the /version endpoint ",
    "location": "tcp://10.96.0.1:443",
    "name": "K8s Version Disclosure",
    "reference": {
      "id": "KHV002",
      "source": "https://aquasecurity.github.io/kube-hunter/kb/KHV002",
    },
    "severity": "MEDIUM",
  },
  {
    "attributes": {
      "evidence": "b'{"kind":"APIVersions","versions":["v1"],"serverAddressByClientCIDRs":[{"clientCIDR":"0.0.0.0/0","serverAddress":"172.17.0.2:6443"}]}\\n'",
      "kubeHunterRule": "API Server Hunter",
    },
    "category": "Information Disclosure",
    "description": " The API Server port is accessible. Depending on your RBAC settings this could expose access to or control of your cluster. ",
    "location": "tcp://10.96.0.1:443",
    "name": "Access to API using service account token",
    "reference": {
      "id": "KHV005",
      "source": "https://aquasecurity.github.io/kube-hunter/kb/KHV005",
    },
    "severity": "MEDIUM",
  },
]
`;
